mmrotate/models/dense_heads/oriented_reppoints_head.py

```diff
@@ -1022,172 +628,366 @@ def loss(self,
-                scores = scores[topk_inds, :]
-
-            pts_pred = points_pred.reshape(-1, self.num_points, 2)
-            pts_pred_offsety = pts_pred[:, :, 0::2]
-            pts_pred_offsetx = pts_pred[:, :, 1::2]
-            pts_pred = torch.cat([pts_pred_offsetx, pts_pred_offsety],
-                                 dim=2).reshape(-1, 2 * self.num_points)
-
-            pts_pos_center = points[:, :2].repeat(1, self.num_points)
-            pts = pts_pred * self.point_strides[level_idx] + pts_pos_center
-
-            polys = min_area_polygons(pts)
-            bboxes = poly2obb(polys, self.version)
-
-            mlvl_bboxes.append(bboxes)
-            mlvl_scores.append(scores)
-
-        mlvl_bboxes = torch.cat(mlvl_bboxes)
-
-        if rescale:
-            mlvl_bboxes[..., :4] /= mlvl_bboxes[..., :4].new_tensor(
-                scale_factor)
-        mlvl_scores = torch.cat(mlvl_scores)
-        if self.use_sigmoid_cls:
-            padding = mlvl_scores.new_zeros(mlvl_scores.shape[0], 1)
-            mlvl_scores = torch.cat([mlvl_scores, padding], dim=1)
-
-        if with_nms:
-            det_bboxes, det_labels = multiclass_nms_rotated(
-                mlvl_bboxes, mlvl_scores, cfg.score_thr, cfg.nms,
-                cfg.max_per_img)
-            return det_bboxes, det_labels
+        batch_size = features.size(0)
+        sampled_features = torch.zeros([
+            pt_locations.shape[0],
+            features.size(1),
+            pt_locations.size(1),
+            pt_locations.size(2)
+        ]).to(pt_locations.device)
+
+        for i in range(batch_size):
+            feature = nn.functional.grid_sample(features[i:i + 1],
+                                                pt_locations[i:i + 1])[0]
+            sampled_features[i] = feature
+
+        return sampled_features,
+
+    def feature_cosine_similarity(self, points_features: Tensor) -> Tensor:
+        """Compute the points features similarity for points-wise correlation.
+
+        Args:
+            points_features (Tensor): sampling point feature with
+                shape (N_pointsets, N_points, C)
+
+        Returns:
+            max_correlation (Tensor): max feature similarity in each point set
+            with shape (N_points_set, N_points, C)
+        """
+
+        mean_points_feats = torch.mean(points_features, dim=1, keepdim=True)
+        norm_pts_feats = torch.norm(
+            points_features, p=2, dim=2).unsqueeze(dim=2).clamp(min=1e-2)
+        norm_mean_pts_feats = torch.norm(
+            mean_points_feats, p=2, dim=2).unsqueeze(dim=2).clamp(min=1e-2)
+
+        unity_points_features = points_features / norm_pts_feats
+        unity_mean_points_feats = mean_points_feats / norm_mean_pts_feats
+
+        cos_similarity = nn.CosineSimilarity(dim=2, eps=1e-6)
+        feats_similarity = 1.0 - cos_similarity(unity_points_features,
+                                                unity_mean_points_feats)
```

Comment:
Use the functional api instead of initializing an nn.Module every time.
```suggestion
        feats_similarity = 1.0 - F.cos_similarity(unity_points_features,
                                                unity_mean_points_feats, dim=2, eps=1e-6)
```